Move and rearrange developer setup instructions.

brainwane opened the pull request:
To consolidate developer documentation on Read The Docs, I turned `README.dev.md` into several topic pages within `/docs/` , and linked to them from the doc index. I also updated links in other documentation to point to the new topic pages.

I'll also submit a pull request to zulip.github.io changing zulip.org links to point to the new topic pages.

Fixes: #669.

smarx commented:
*Automated message from Dropbox CLA bot*

@brainwane, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented:
Great, thanks for working on this @brainwane!  A few comments:
* The tests are failing due to some lint issues (newline at end of file)
* We should leave a README.dev.md file placeholder since we can't update zulip.org atomically
* Here are some proposed changes to the filenames:

  - docs/install-docker-experimental-dev.md -> docs/install-docker-dev.md (experimental probably doesn't belong in the filename, since I don't want to need to rename it when we make it non-experimental)
  - docs/install-manually-unix-based-dev.md -> docs/install-generic-unix-dev.md
  - docs/install-ubuntu-1404-trusty-without-vagrant-dev.md -> docs/install-ubuntu-without-vagrant-dev.md (16.04 basically works here now so it'll soon be supported on that page)
  - docs/test-suite-dev.md perhaps should just get merged into a first section in testing.md, if it's going to be separated from the rest.  I'm willing for fixing this to come in a follow-up PR to get this merged soon (and thus avoid merge conflicts).

* Have you tested every link in the moved files?  Regardless probably worth doing again after doing the renames.

Otherwise this looks good (though probably there will be)

brainwane commented:
@timabbott Ready to re-review. Thanks for the critique. I believe I've addressed everything. And I ran `make html` and used `git grep` as well as visual inspection to test all links after the changes.

timabbott commented:
Cool, a few issues I notice:
* testing.html appears twice in the index; let's drop the place in the "installation guide" where it appears
* The link at the end of the "first-time contributor guide" to the "using" guide is broken
* The installation instructions section takes up a lot of space in the left sidebar index; maybe we should move it to the bottom?

timabbott commented:
Also, I think dev-install-choices.html should perhaps be renamed to be a high-level page on "Zulip development environment" and the current links can be a section on installation, and maybe then the second half can be the "Using the development environment" page plus links on to testing, the tutorials, and other things in the code contribution guide someone might want to read next.  That seems like a still reasonably short page that would feel like a good overview.

brainwane commented:
@timabbott, I believe I've addressed most of your points in the new commits in the branch.

> The detailed installation instructions sections take up a lot of space in the left sidebar index; maybe we should move it to the bottom so the things that current developers are using often are highlighted better?

When you say that the detailed installation instructions sections take up a lot of space in the lefthand navbar, I'm not sure I'm seeing what you're seeing, because on my machine, all the headers for page sections start out minimized.

I also agree with you that there's likely a better possible arrangement of the installation, usage, testing, and tutorial-type guides. And I agree that rearranging those suites or sequences is something that doesn't need to block this pull request.

In "Remove duplicate testing link in TOC. (a295109)" I may have done something suboptimal regarding embedding/linking to a nonlocal image, but this was necessary to make the HTML build properly on my machine. Please feel free to fiddle with it if there are different affordances possible on RTD.

timabbott commented:
Here's a screenshot of what I meant regarding the TOC:
![image](https://cloud.githubusercontent.com/assets/2746074/16790874/360ceeee-486e-11e6-8854-
d0063578f00b.png)

timabbott commented:
The comment I made on breaking up the commits for README.prod.md applies here as well; I'd love to have the first section be done more incrementally if the rebase is reasonably quick for you.  Also, fixups like `Remove duplicate testing link in TOC.` should be squashed into the commit that introduced the issue.

brainwane commented:
I see my email reply didn't come through :(

I'll take a look tomorrow to see how quickly I can rebase to split this up. And I mostly have squashed fixups like that into the relevant revisions -- in this case I just wanted to err on the side of ensuring you saw it was addressed. Will squash that & the similar ones.

brainwane commented:
@timabbott You said "I'd love to have the first section be done more incrementally if the rebase is reasonably quick for you". Turns out it's not. I think it'll take me another hour; I'll finish it in the morning unless you say I shouldn't. However, I have squashed the fixup commits, and I'll rearrange the table of contents so the less-used installation options don't take up as much vertical space in the sidebar.

brainwane commented:
@timabbott This is ready to review. It leaves README.dev.md as a sort of table of contents. In my next commits I'll reduce how much room the lists of installation pages take up in the left-hand RTD navbar, and turn the remnant of README.dev.md into a better arrangement per [your thinking a few days ago](https://github.com/zulip/zulip/pull/1218#issuecomment-232152548). But I think this is merge-ready.

timabbott commented:
OK, great, merged (after one fix to prevent "Possible testing issues" being a top-level heading)!  Definitely worth doing those follow-up projects since I think the new state definitely has some rough edges around those things, but they can be a new PR.  The other follow-up thing we should add to the TODO list is that I think `test-suite-dev.md` should probably be merged into testing.md -- I don't think we need a separate page for it.

Thanks @brainwane!

timabbott commented:
(And sorry for the slow reply on this one -- I was away for the weekend with an extended wedding)
